### PR TITLE
Define VM type and its behavior

### PR DESCRIPTION
**Pull request by gouyang**
Signed-off-by: Guohua Ouyang <gouyang@redhat.com>

### CODE REVIEW

**Comment by gouyang** [comment]
@nellyc @lukas-bednar @gbenhaim Please help review this because I want it to be merged asap, so I can refactor other tests based on it.

**Comment by lukas-bednar** [comment]
it looks good to me ... I just wonder if such helpers might be in kubevirt repository already ?

In addition I see this error in logs:
```
Error: Failed to compile tests
```

**Comment by gouyang** [comment]
> it looks good to me ... I just wonder if such helpers might be in kubevirt repository already ?
It seems no, but it doesn't matter as these helpers are implemented by `virtctl` and `oc`.

> In addition I see this error in logs:

Fixed.

**Comment by gbenhaim** [comment]
@shiywang Can you please review?

File: tests/types.go [hunk 1]
```diff
@@ -1,8 +1,91 @@
...
+	Name              string
+	Type              string
+	Manifest          string
+	TemplateInCluster string
+	TemplateFromFile  string
+	TemplateParams    []string
+}
+
+func (vm VirtualMachine) Create() (string, error) {
+	args := []string{"create", "-n", NamespaceTestDefault, "-f", vm.Manifest}
```
Comment (lukas-bednar) [inline tests/types.go]: `s/NamespaceTestDefault/ktests.NamespaceTestDefault/`

File: tests/util.go [hunk 2]
```diff
@@ -29,7 +29,10 @@ type Result struct {
 	params        []string
 }
 
-const NamespaceTestDefault = "kubevirt-test-default"
+const (
+	NamespaceTestDefault  = "kubevirt-test-default"
```
Comment (lukas-bednar) [inline tests/util.go]: This constant seems familiar to me, it is defined in kubevirt repo.

**Comment by gouyang** [comment]
I'm fine to just use the constant from `kubevirt`, change soon.

File: tests/util.go (hunk 2 above)
Comment (lukas-bednar) [reply tests/util.go]: Please remove this one.

Comment (gouyang) [reply tests/util.go]: this one is not introduced in this PR, I'm afraid other tests are using it. It need an overhaul to remove it.

Comment (lukas-bednar) [reply tests/util.go]: Hm ... I see now ...

**Comment by gbenhaim** [comment]
@lukas-bednar are you ok with this PR?

File: tests/types.go [hunk 3]
```diff
@@ -1,8 +1,90 @@
...
+	Type              string
+	Manifest          string
+	TemplateInCluster string
+	TemplateFromFile  string
+	TemplateParams    []string
+}
+
+func (vm VirtualMachine) Create() (string, error) {
+	args := []string{"create", "-n", ktests.NamespaceTestDefault, "-f", vm.Manifest}
+	return ktests.RunCommand("oc", args...)
```
Comment (lukas-bednar) [inline tests/types.go]: I would replace `oc` with `ktests.KubeVirtOcPath`

**Comment by gouyang** [comment]
@lukas-bednar @gbenhaim Please review again, I think it can be merged.

**Comment by gouyang** [comment]
ci test please

File: tests/framework/types.go [hunk 4]
```diff
@@ -23,3 +24,96 @@ const (
 	ShortTimeout = time.Duration(2) * time.Minute
 	LongTimeout  = time.Duration(4) * time.Minute
 )
+
+// VirtualMachine can be a vm, vmi, vmirs, vmiPreset.
+type VirtualMachine struct {
```
[the remark this answers is not in the document]
Comment (gouyang) [reply tests/framework/types.go]: @shiywang do you have a strong reason on that? To my knowledge, it's the best practice to put them here.

File: tests/framework/types.go [hunk 5]
```diff
@@ -23,3 +24,96 @@ const (
...
+	Type              string
+	UID               string
+	Manifest          string
+	TemplateInCluster string
+	TemplateFromFile  string
+	TemplateParams    []string
+}
+
+func (vm VirtualMachine) Create() (string, string, error) {
+	args := []string{"create", "-n", ktests.NamespaceTestDefault, "-f", vm.Manifest}
```
Comment (shiywang) [inline tests/framework/types.go]: duplicate `"-n", ktests.NamespaceTestDefault` with RunCommand()

**Pull request merged**